### PR TITLE
cmake,UI: Fix SOVERSION on Linux

### LINKED ISSUE #10737: Trying to build Flatpak from Git , `libobs-opengl.so` not found
### Operating System Info

Other

### Other OS

Fedora 40

### OBS Studio Version

Git

### OBS Studio Version (Other)

_No response_

### OBS Studio Log URL

https://gist.github.com/AFCMS/ce8400fcc09288c00478cf4bc2ea95c2

(not a obsproject.com uploaded log file because the interface didn't show up at all)

### OBS Studio Crash Log URL

_No response_

### Expected Behavior

I would like the software to start and work.

### Current Behavior

I wanted to build the latest Git version of OBS.

For that I cloned recursively the Git repo and executed the following:
```shell
flatpak-builder --user --install build-dir --force-clean ./build-aux/com.obsproject.Studio.json
```

The build worked fine but when I run `flatpak run --user com.obsproject.Studio` (no other versions of OBS are installed in Flatpak) in my terminal a small popup appear with `Failed to initialize video:  Graphics module not found`, while in logs there are two error lines:
```
error: os_dlopen(libobs-opengl.so.30->libobs-opengl.so.30): libobs-opengl.so.30: Ne peut ouvrir le fichier d'objet partagé: Aucun fichier ou dossier de ce type
error: Failed to initialize video:  Graphics module not found
```

image

### Steps to Reproduce

I am unsure, really. I would have expected the Flatpak to work straight since it's a cross-platform format and the build process went smoothly.

I did multiple tests on my hardware/config and all lead to the same result.

### Anything else we should know?

_No response_

### PR DESCRIPTION
**Pull request by tytan652**
### Description

Fixes #10737
Supersedes #10739

SOVERSION is either the major number of a semver or an arbitrary number incremented at each API/ABI break.

### Motivation and Context

CMake 3 does not apply RPATH like CMake 2 so it unraveled this issue.

We didn't released a CMake 3.0 Linux build, so no breaking change.

### How Has This Been Tested?

Build and installed the Flatpak locally and OBS finds `libobs-opengl.so.30`.

Non-Flatpak local build seems to work too.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: UI/CMakeLists.txt [hunk 1]
```diff
@@ -97,8 +97,11 @@ endif()
 foreach(graphics_library IN ITEMS opengl metal d3d11)
   string(TOUPPER ${graphics_library} graphics_library_U)
   if(TARGET OBS::libobs-${graphics_library})
-    target_compile_definitions(obs-studio
-                               PRIVATE DL_${graphics_library_U}="$<TARGET_FILE_NAME:OBS::libobs-${graphics_library}>")
+    target_compile_definitions(
+      obs-studio
+      PRIVATE
+        DL_${graphics_library_U}="$<$<IF:$<BOOL:${OS_LINUX}>,TARGET_SONAME_FILE_NAME,TARGET_FILE_NAME>:OBS::libobs-${graphics_library}>"
```
Comment (PatTheMav) [inline UI/CMakeLists.txt:103]: Is this change necessary to get the _versioned_ name of the library set as a define?

Comment (tytan652) [reply UI/CMakeLists.txt:103]: To get the SONAME one only on Linux yes. `TARGET_FILE_NAME` will put `libobs-opengl.so.30.1.0` where we need the SONAME (`libobs-opengl.so.30`) given by `TARGET_SONAME_FILE_NAME`.

Otherwise, we just create the same breakage we filename reversed.

**Review by AFCMS — COMMENTED** [review]
Tested, works too

File: UI/CMakeLists.txt (hunk 1 above)
Comment (PatTheMav) [reply UI/CMakeLists.txt:103]: I'd check if just using `TARGET_SONAME_FILE_NAME` for all platforms shouldn't work out of the box - we set `NO_SONAME` for macOS and Windows doesn't support that anyway, so it would be interesting to see if that generator expression falls back on the target name then.

Otherwise I'd be fine using a conditional branch here instead (you can compose a string to _contain_ a generator expression, it doesn't need to be used at the immediate call site).

Comment (tytan652) [reply UI/CMakeLists.txt:103]: > I'd check if just using `TARGET_SONAME_FILE_NAME` for all platforms shouldn't work out of the box - we set `NO_SONAME` for macOS and Windows doesn't support that anyway, so it would be interesting to see if that generator expression falls back on the target name then.

I'm only fixing the Linux path, I don't plan to test CMake behavior on other platform.

> Otherwise I'd be fine using a conditional branch here instead (you can compose a string to _contain_ a generator expression, it doesn't need to be used at the immediate call site).

I preferred to keep it simple and avoid duplicating code and/or effort in the three platform file.

Comment (PatTheMav) [reply UI/CMakeLists.txt:103]: We cannot just fix a thing for one platform, we always have to code with _all_ supported platforms in mind, particularly for code that effects all platforms (in this case `libobs-opengl` as it's not a platform-specific target).

Comment (tytan652) [reply UI/CMakeLists.txt]: Changed the logic to use `PLATFORM_ID`

**Review by AFCMS — COMMENTED** [review]
Still works for me
